overlord/snapstate/storehelpers.go

```diff
@@ -867,40 +721,194 @@ func refreshCandidatesCore(ctx context.Context, st *state.State, names []string,
 			userID = fallbackID
 		}
 		actionsByUserID[userID] = append(actionsByUserID[userID], action)
-		if snapst.IgnoreValidation {
-			ignoreValidationByInstanceName[installed.InstanceName] = true
-		}
-		nCands++
+
 		return nil
 	}
 
-	holds, err := SnapHolds(st, names)
+	// TODO: is this right? why do we only pass in the requested names here?
+	// what about when we are refreshing all snaps?
+	holds, err := SnapHolds(st, plan.requested)
+	if err != nil {
+		return updatePlan{}, err
+	}
+
+	// determine current snaps and create actions for each snap that needs to
+	// be refreshed
+	current, err := collectCurrentSnaps(all, holds, addCand)
```

Comment:
For the TODO, this is consistent with what we do for on master: https://github.com/snapcore/snapd/blob/master/overlord/snapstate/storehelpers.go?plain=1#L877

`names` there is equivalent to the set of snaps that we were requested to refresh, so it could be an empty slice for auto-refreshes/general refreshes.